SU2_CFD/src/solvers/CSpeciesSolver.cpp

```diff
@@ -360,39 +362,66 @@ void CSpeciesSolver::BC_Inlet(CGeometry* geometry, CSolver** solver_container, C
 
       su2double Normal[MAXNDIM] = {0.0};
       for (auto iDim = 0u; iDim < nDim; iDim++) Normal[iDim] = -geometry->vertex[val_marker][iVertex]->GetNormal(iDim);
-      conv_numerics->SetNormal(Normal);
 
-      /*--- Allocate the value at the inlet ---*/
+      if(bounded_scalar){
 
-      auto V_inlet = solver_container[FLOW_SOL]->GetCharacPrimVar(val_marker, iVertex);
+        /*--- Obtain flow velocity vector at inlet boundary node ---*/
+        su2double Velocity_Inlet[MAXNDIM] = {0.0}, density_inlet, mflux_inlet;
+        for (auto iDim = 0u; iDim < nDim; iDim++) Velocity_Inlet[iDim] = solver_container[FLOW_SOL]->GetNodes()->GetVelocity(iPoint, iDim);
+        
+        /*--- Obtain density at inlet boundary node ---*/
+        density_inlet = solver_container[FLOW_SOL]->GetNodes()->GetDensity(iPoint);
 
-      /*--- Retrieve solution at the farfield boundary node ---*/
+        /*--- Compute mass flux on current node ---*/
+        mflux_inlet = density_inlet * GeometryToolbox::DotProduct(nDim, Velocity_Inlet, Normal);
```

Comment:
Would it be possible to set this in the numerics class instead of updating the residual manually? Because this is not doing the right upwind treatment in case there is reversed flow.